UserOfficeProject/user-office-backend
# Stfc user office project 200 generic templates

## Description


Closes: https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

Note: When downloading a submitted proposal, responses submitted using the generic template feature does not appear in pdf correctly, issue opened for this here: https://github.com/UserOfficeProject/stfc-user-office-project/issues/222

Replicated Sample Declaration functionality, in the form of Generic Templates. When logged in as a user officer, they can create as many templates as they want, and each are independent from each other. For example, using this feature they can create a `Grants` template and also `Publications` template and many more templates which they want to create.  

When logged in as a user, the user can fill out their answers to the generic templates created. The added feature allows them to input as many responses as they want to add (within the limits set by the user officer) using the `Add` button. The user can `Clone` and `Delete` the added response.

![image](https://user-images.githubusercontent.com/72739242/136419912-ee79e042-d28c-400c-bd3d-87597e269a05.png)
What user sees when submitting a proposal.

![image](https://user-images.githubusercontent.com/72739242/136420001-72a86308-1839-4f28-aae6-90b64dc62ba2.png)
What user sees after clicking the `Add` button to add a `Grant`

![image](https://user-images.githubusercontent.com/72739242/136420188-77b88210-2adf-4627-9f39-73067644f1b1.png)
When logged in as a user officer, they can create new templates and edit existing ones

![image](https://user-images.githubusercontent.com/72739242/136420597-b4d52a66-735c-4d4e-83f9-21f9f2da2fdf.png)
Selecting `Add Sub Template` will allow a created generic template to be added into the questionnaire
## Motivation and Context


Allows customised questions to be asked in a straight forward process. Also, allows user to add, clone, delete their responses freely. 
https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

## How Has This Been Tested





- Browser testing
   - Checked if process implemented was flowing correctly.
   - Logged in as a user officer, navigated to `Sub Templates` and created a template and added it to the proposal questionnaire.
   - Logged in as a user and filled out the questionnaire to check if everything was working as expected. 

- Backend tests
   - Wrote backend tests and ran them to see if they passed, if they didn't went through and made appropriate changes to the code to ensure the tests passed.

## Fixes


Fixes: https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

## Changes



- Added DB Patch to create a `generic_templates` table, add `GENERIC_TEMPLATE` as a Category and question data type and add `GENERIC_TEMPLATE_BASIS` as a question data type
- Added GenericTemplate references in `dependencyConfigs`
- Added `GenericTemplateDataSource` files
- Added `GenericTemplate Authorisation`
- Added `GenericTemplateMutations`
- Added references to GenericTemplates in files `proposal.ts, records.ts, index.ts, QuestionaryAuthorization.ts, QuestionRegistry.ts, Template.ts `
- Added Generic Template `queries`
- Added GenericTemplate type file and in relavent `fieldConfig.ts` files. 
- Added GenericTemplate references in `resolvers` folder 

## Depends on


Frontend PR: https://github.com/UserOfficeProject/user-office-frontend/pull/580


## Tests included/Docs Updated?



- [ x ] I have added tests to cover my changes.
- [ x ] All relevant doc has been updated
